pkg/operator/connectivitycheckcontroller/connectivity_check_controller.go

```diff
@@ -145,56 +146,70 @@ func (c *connectivityCheckController) Sync(ctx context.Context, syncContext fact
+		for _, existingCheck := range existingChecks.Items {
+			err := c.operatorcontrolplaneClient.ControlplaneV1alpha1().PodNetworkConnectivityChecks(c.namespace).Delete(ctx, existingCheck.Name, metav1.DeleteOptions{})
 			if err != nil {
 				return err
 			}
 		}
 		return nil
 	}
 
-	checks, err := c.podNetworkConnectivityCheckFn(ctx, syncContext)
+	newChecks, err := c.podNetworkConnectivityCheckFn(ctx, syncContext)
 	if err != nil {
 		return err
 	}
 
 	pnccClient := c.operatorcontrolplaneClient.ControlplaneV1alpha1().PodNetworkConnectivityChecks(c.namespace)
-	for _, check := range checks {
-		existing, err := pnccClient.Get(ctx, check.Name, metav1.GetOptions{})
+	newCheckNames := sets.NewString()
+	for _, newCheck := range newChecks {
+		newCheckNames.Insert(newCheck.Name)
+		existing, err := pnccClient.Get(ctx, newCheck.Name, metav1.GetOptions{})
 		if err == nil {
-			if equality.Semantic.DeepEqual(existing.Spec, check.Spec) {
+			if equality.Semantic.DeepEqual(existing.Spec, newCheck.Spec) {
 				// already exists, no changes, skip
 				continue
 			}
 			updated := existing.DeepCopy()
-			updated.Spec = *check.Spec.DeepCopy()
+			updated.Spec = *newCheck.Spec.DeepCopy()
 			_, err := pnccClient.Update(ctx, updated, metav1.UpdateOptions{})
 			if err != nil {
-				syncContext.Recorder().Warningf("EndpointDetectionFailure", "%s: %v", resourcehelper.FormatResourceForCLIWithNamespace(check), err)
+				syncContext.Recorder().Warningf("EndpointDetectionFailure", "%s: %v", resourcehelper.FormatResourceForCLIWithNamespace(newCheck), err)
 				continue
 			}
-			syncContext.Recorder().Eventf("EndpointCheckUpdated", "Updated %s because it changed.", resourcehelper.FormatResourceForCLIWithNamespace(check))
+			syncContext.Recorder().Eventf("EndpointCheckUpdated", "Updated %s because it changed.", resourcehelper.FormatResourceForCLIWithNamespace(newCheck))
 		}
 		if errors.IsNotFound(err) {
-			_, err = pnccClient.Create(ctx, check, metav1.CreateOptions{})
+			_, err = pnccClient.Create(ctx, newCheck, metav1.CreateOptions{})
 		}
 		if err != nil {
-			syncContext.Recorder().Warningf("EndpointDetectionFailure", "%s: %v", resourcehelper.FormatResourceForCLIWithNamespace(check), err)
+			syncContext.Recorder().Warningf("EndpointDetectionFailure", "%s: %v", resourcehelper.FormatResourceForCLIWithNamespace(newCheck), err)
 			continue
 		}
-		syncContext.Recorder().Eventf("EndpointCheckCreated", "Created %s because it was missing.", resourcehelper.FormatResourceForCLIWithNamespace(check))
+		syncContext.Recorder().Eventf("EndpointCheckCreated", "Created %s because it was missing.", resourcehelper.FormatResourceForCLIWithNamespace(newCheck))
 	}
 
 	// TODO for checks which longer exist, mark them as completed
 
-	// TODO reap old connectivity checks
+	// reap old connectivity checks
+	for _, existingCheck := range existingChecks.Items {
```

Comment:
This prevents an external actor from creating a new connectivity check and managing it differently. I suggest adding an annotation or label to indicate, "managed by this operator" and then only deleting those items managed by this operator.

Since this is an alpha API, I'll allow an upgrade that leaves old content around or deletes old content as you wish.  But to delete old content, you'll need a two stage migration where the first stage labels everything in the namespace in 4.n-1 .

Reply:
sure, now this would update old check objects with `managedBy` label, but this won't delete old stale objects during migration. hope that's ok.